File: Code/C++/871. Minimum Number of Refueling Stops.cpp
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    int minRefuelStops(int target, int startFuel, vector<vector<int>>& stations) {
        const int n = stations.size();
        priority_queue<int> pq;
        int res=0;
        int i=0;
        while (startFuel < target){
            while (i < n && stations[i][0] <= startFuel){
                pq.push(stations[i][1]);
                i++;
            }
            if (pq.empty()){
                return -1;
            }
            startFuel += pq.top();
            pq.pop();
            res++;
        }
        return res;

    }

};

File: Code/Python/871. Minimum Number of Refueling Stops.py
from heapq import *

class Solution:
    def minRefuelStops(self, target, startFuel, stations):
        start_fuel = startFuel
        pq = []
        res = 0
        idx = 0
        n = len(stations)
        while start_fuel < target:
            while idx < n and stations[idx][0] <= start_fuel:
                heappush(pq, -stations[idx][1])
                idx += 1
            if not pq:
                return -1
            
            start_fuel += -heappush(pq)
            res += 1

        return res
        